src/Scope/TreeTraverser.php

```diff
@@ -0,0 +1,193 @@
+<?php
+declare(strict_types=1);
+
+namespace LanguageServer\Scope;
+
+use LanguageServer\DefinitionResolver;
+use Microsoft\PhpParser\ClassLike;
+use Microsoft\PhpParser\FunctionLike;
+use Microsoft\PhpParser\MissingToken;
+use Microsoft\PhpParser\Node;
+use Microsoft\PhpParser\Node\Expression;
+use Microsoft\PhpParser\Token;
+use Microsoft\PhpParser\TokenKind;
+use phpDocumentor\Reflection\Fqsen;
+use phpDocumentor\Reflection\Types;
+
+/**
+ * Traversers AST with Scope information.
+ */
+class TreeTraverser
```

Comment:
The parser already exposes utilities to traverse the tree (and uses the iterator pattern instead of the visitor pattern, which is superior imo). Given that this is a fair amount of code, could you explain why this is needed? What is the difference between this scope-tree traverser and a general traverser?

Reply:
I need to look into the parser code and see if we could use it - my initial guess is that we can't though, since the traversing needs to happen in the following fashion:

```
Scope = []                      //    Scope 1 - Root
    Scope = [$this]             //    Scope 2 - Class entered 
        Scope = [$this, $a]     //    Scope 3 - Function entered
    Scope = [$this]             // !! Scope 2 - Same scope again 
Scope = []                      // !! Scope 1 - Same scope again
```

The lines marked with `!!` - if we only get a callback per `Node` or `Token`, we cannot know when a scope is exited. Additionally, we'd need to manage a stack manually instead of relying on the call stack.

Further, we need information on the context of the element - if we have a `BracedStatementList` (or whatever it's called), we need to know if it is a `functionBodyOrSemicolon` to start a new `Scope`, so the `Node` alone is not sufficient.

If the parser does not provide a traverser that fits those two needs, I think that working around them might make the code more confusing.

Reply:
I looked into this, and  `getDescendantNodesAndTokens` or `getDescendantNodes` cannot really be used, since we won't know when a node is exited and the scope should be restored.

I think that `TreeTraverser` could be converted into a generator yielding something - e.g. rename `Scope` to `TraversingContext` which contains both `$node` and `$variables` etc. Do you think that's worth pursuing?

Reply:
I looked into this, and `getDescendantNodesAndTokens` or `getDescendantNodes` cannot really be used, since we won't know when a node is exited and the scope should be restored.

I think that `TreeTraverser` could be converted into a generator yielding something - e.g. rename `Scope` to `TraversingContext` which contains both `$node` and `$variables` etc. Do you think that's worth pursuing?

(I accidentally posted this with a wrong GitHub account first - sorry about the noise)

Reply:
So first of all I still think parent traversal is the better way to go.
And then `getDescendentNodes` afaik takes a callback to decide whether a node should be entered or not.
For traversal that needs full control over the recursive aspect I did a PR a while ago to tolerant-php-parser with RecursiveIterator support: https://github.com/Microsoft/tolerant-php-parser/pull/139
I didn't need it at the time though and it turned out it was a tad slower than the generators. It would be interesting to see how it compares to this TreeTraverser and the previous implementation, especially on latest PHP 7.2.

Reply:
The `RecursiveIterator` thing certainly looks interesting. It would require manual handling of a stack though, if I understand it correctly.

I think that traversing the code backwards to find types has a few problems, which I tried to solve with this:

#### 1. Duplicated work:

Find references on line 2:
```php
1 $a = new A;     // 2.                    Ok, `$a is A`
2 $b = $a->foo(); // 1. Start traversing - need to find what $a is
3 $b->bar();
```
After that, find references on line 3:
```php
1 $a = new A;     // 3. Duplicated work  - Ok, `$a is A`
2 $b = $a->foo(); // 2. Duplicated work  - Need to find out what $a is
3 $b->bar();      // 1. Start Traversing - need to find what $b is to get reference to bar
```
This could of course be solved with some sort of memoization. It would probably end up building a scope, but in reverse order.

#### 2. Slowness of parsing backwards in Tolerant PHP Parser
Getting a previous sibling gets its parent, enumerates its children, and when it finds the original node, it returns the previously met one.

#### (3. Harder to understand - this is mostly a matter of preference maybe)
Since code is evaluated from top to bottom, an approach evaluating code from top to bottom is easier to understand, at least for me.

*I'm going to sleep now, so I won't reply for a while. I might take a look at the other PRs and comments tomorrow.*

Reply:
> 1. Duplicated work:

I don't think this problem has anything to do with how to traverse. The difference is that this PR introduces a Scope object, where before this was always computed JIT. I never had any performance problems with find-refs on variables though because their scope is naturally small in PHP, so it's a CPU vs memory usage trade off.

> 2. Slowness of parsing backwards in Tolerant PHP Parser
> Getting a previous sibling gets its parent, enumerates its children, and when it finds the original node, it returns the previously met one.

This is a good point. With PHPParser I added the siblings as properties so that was fast. We could do the same for tolerant-parser.

----------

Are you maintaining the same behaviour as before with the new traverser?

Reply:
>  I never had any performance problems with find-refs on variables though because their scope is naturally small in PHP, so it's a CPU vs memory usage trade off.

It might be that the performance benefits here come from caching `getResolvedName` results in the `Scope` object. I started working on this because XDebug profiling showed finding variable types as a significant cost. XDebug has very high overhead though, so it might be completely inaccurate. It might be worth trying to cache the `getResolvedName` results alone, especially if you don't like the `Scope` stuff.

> Are you maintaining the same behaviour as before with the new traverser?

I believe so - the variables in `Scope` are overwritten when a new assignment is met, so on line 3 `$a` would be a string.